### PR TITLE
Schem file limits

### PR DESCRIPTION
**Pull request by Ivan8or**
## Overview

This pull request adds a way to set limits on per-player schematics - adding config options which define the maximum space a player's schematics folder can take up (in kb) as well as the maximum number of schematics a player can have.

If there were to continue to be no limit on schematic size, players could conceivably run a script to save a large clipboard under multiple schematic names and in fill up the server's disk space.

**Fixes https://github.com/IntellectualSites/FastAsyncWorldEdit/issues/742**

## Description

I have added some rough but functional logic which ensures that while the config options are set, players must have the required room (filesize & slots) in order to save their schematics.

I have also made some adjustments to what `/schematic list` displays, now also showing the size of each schematic on the disk, as well as the total size of all combined schematics (as well as the size quota, if that permission is set)

## Checklist

- [x] I included all information required in the sections above
- [x] I tested my changes and approved their functionality
- [x] I ensured my changes do not break other parts of the code
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/FastAsyncWorldEdit/blob/1.16/CONTRIBUTING.md)

### CODE REVIEW

**Review by N0tMyFaultOG — CHANGES_REQUESTED** [review]
Checkstyle.

File: worldedit-core/src/main/java/com/sk89q/worldedit/util/formatting/component/MessageBox.java [hunk 1]
```diff
@@ -33,7 +33,12 @@
  */
 public class MessageBox extends TextComponentProducer {
 
-    private static final int GUARANTEED_NO_WRAP_CHAT_PAGE_WIDTH = 47;
+    //decremented by one (from 47) to account for minecraft font characters having varying widths
+
+    // method for applying border needs to be replaced in the future  - need to
+    // be able to measure length of title (and prefix) by pixel width, not by character width
+    // (to avoid wrapping in case of thick chars, or not enough spacers in case of thin chars)
+    private static final int GUARANTEED_NO_WRAP_CHAT_PAGE_WIDTH = 46;
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/util/formatting/component/MessageBox.java:41]: I'd like to keep full compatibility here with WE. If this is an actual issue, it should be addressed in WE instead.

**Review by SirYwell — CHANGES_REQUESTED** [review]

File: worldedit-core/src/main/java/com/sk89q/worldedit/util/formatting/component/MessageBox.java (hunk 1 above)
Comment (Ivan8or) [reply worldedit-core/src/main/java/com/sk89q/worldedit/util/formatting/component/MessageBox.java:41]: alright, i'll see what I can do

**Comment by Ivan8or** [comment]
> Checkstyle.

what's that o.o

**Comment by N0tMyFaultOG** [comment]
> > Checkstyle.
> 
> what's that o.o

What you apply to your IDE so your code contribution follows our standards and does not look out of place, see https://github.com/IntellectualSites/FastAsyncWorldEdit/tree/main/config/checkstyle

**Comment by Ivan8or** [comment]
> > > Checkstyle.
> > 
> > 
> > what's that o.o
> 
> What you apply to your IDE so your code contribution follows our standards and does not look out of place, see https://github.com/IntellectualSites/FastAsyncWorldEdit/tree/main/config/checkstyle
> 
> >  GitHub**[IntellectualSites/FastAsyncWorldEdit](https://github.com/IntellectualSites/FastAsyncWorldEdit)**Blazingly fast world manipulation for artists, builders and everyone else: https://www.spigotmc.org/resources/13932/ - IntellectualSites/FastAsyncWorldEdit

aah gotcha!

I should have those same settings on my IDE since I cloned the project, right?

**Comment by N0tMyFaultOG** [comment]
> > > > Checkstyle.
> > > 
> > > 
> > > what's that o.o
> > 
> > 
> > What you apply to your IDE so your code contribution follows our standards and does not look out of place, see https://github.com/IntellectualSites/FastAsyncWorldEdit/tree/main/config/checkstyle
> > >  GitHub**[IntellectualSites/FastAsyncWorldEdit](https://github.com/IntellectualSites/FastAsyncWorldEdit)**Blazingly fast world manipulation for artists, builders and everyone else: https://www.spigotmc.org/resources/13932/ - IntellectualSites/FastAsyncWorldEdit
> 
> aah gotcha!
> 
> I should have those same settings on my IDE since I cloned the project, right?
> 
> >  GitHub**[IntellectualSites/FastAsyncWorldEdit](https://github.com/IntellectualSites/FastAsyncWorldEdit)**Blazingly fast world manipulation for artists, builders and everyone else: https://www.spigotmc.org/resources/13932/ - IntellectualSites/FastAsyncWorldEdit
> 
> >  SpigotMC - High Performance Minecraft**[Fast Async WorldEdit](https://www.spigotmc.org/resources/fast-async-worldedit.13932/)**Blazingly fast world manipulation for builders, large networks and developers.

Yeah, you have the folder as well. Using the checkstyle prevents adding more code being out of place/needing a huge reformat, like we did the past months (see the 30k diff PRs for reference).

**Comment by Ivan8or** [comment]
> Yeah, you have the folder as well. Using the checkstyle prevents adding more code being out of place/needing a huge reformat, like we did the past months (see the 30k diff PRs for reference).

I just ran a scan with the included checkstyle and it gave me a bunch of warnings across the entire project haha

are there certain warnings I should look out for, while ignoring others? A lot of them don't even make sense to me, like `CRLF and CR line endings are prohibited, but this file uses them`

File: worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java [hunk 2]
```diff
@@ -394,6 +394,18 @@
             " - Requires combine_stages = true"
         })
         public boolean SEND_BEFORE_HISTORY = false;
+
+        @Comment({
+                "Sets a maximum limit (in kb) for the size of a player's schematics directory (per-player mode only)",
+                "set to -1 to have no limit"
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java:400]:
```suggestion
                "Set to -1 to have no limit"
```

Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java:400]:
```suggestion
                "Set to -1 to have no limit"
```

File: worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java [hunk 3]
```diff
@@ -394,6 +394,18 @@
...
+        })
+        public int PERPLAYER_FILESIZELIMIT = -1;
+
+        @Comment({
+                "Sets a maximum limit for the amount of schematics in a player's schematics directory (per-player mode only)",
+                "set to -1 to have no limit"
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java:406]:
```suggestion
                "Set to -1 to have no limit"
```

File: worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java [hunk 4]
```diff
@@ -394,6 +394,18 @@
             " - Requires combine_stages = true"
         })
         public boolean SEND_BEFORE_HISTORY = false;
+
+        @Comment({
+                "Sets a maximum limit (in kb) for the size of a player's schematics directory (per-player mode only)",
+                "set to -1 to have no limit"
+        })
+        public int PERPLAYER_FILESIZELIMIT = -1;
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java:402]:
```suggestion
        public int PER_PLAYER_FILE_SIZE_LIMIT = -1;
```

File: worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java [hunk 5]
```diff
@@ -394,6 +394,18 @@
...
+
+        @Comment({
+                "Sets a maximum limit for the amount of schematics in a player's schematics directory (per-player mode only)",
+                "set to -1 to have no limit"
+        })
+        public int PERPLAYER_FILENUMLIMIT = -1;
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/boydti/fawe/config/Settings.java:408]:
```suggestion
        public int PER_PLAYER_FILE_NUM_LIMIT = -1;
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 
 package com.sk89q.worldedit.command;
 
+import javax.annotation.Nullable;
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:22]: Imports

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 7]
```diff
@@ -559,16 +586,100 @@ public void list(Actor actor, LocalSession session,
...
+                }
+
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        String header_bytes_elem = String.format("%.1fkb", total_bytes / 1000.0);
+
+        if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PERPLAYER_FILESIZELIMIT > -1) {
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:619]:
```suggestion
        if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_SIZE_LIMIT > -1) {
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 8]
```diff
@@ -559,16 +586,100 @@ public void list(Actor actor, LocalSession session,
...
+            header_bytes_elem += String.format(" / %dkb",
+                Settings.IMP.EXPERIMENTAL.PERPLAYER_FILESIZELIMIT);
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:621]:
```suggestion
                Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_SIZE_LIMIT );
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 9]
```diff
@@ -559,16 +586,100 @@ public void list(Actor actor, LocalSession session,
...
+                }
                 return msg.create();
             });
-        PaginationBox paginationBox = PaginationBox.fromComponents("Available schematics", pageCommand, components);
+
+        long total_bytes = 0;
+        File parent_dir = new File(dir.getAbsolutePath() + (playerFolder ? File.separator + uuid.toString() : ""));
+        try {
+            for (File schem : parent_dir.listFiles()) {
+
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:607]:
```suggestion
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 10]
```diff
@@ -611,6 +722,58 @@ public Void call() throws Exception {
...
+                    int iter = 1;
+                    while (new File(overwrite_path + "." + iter + "." + format.getPrimaryFileExtension()).exists()) {
+                        iter++;
+                    }
+                    file = new File(overwrite_path + "." + iter + "." + format.getPrimaryFileExtension());
+                }
+            }
+
+
+            if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PERPLAYER_FILENUMLIMIT > -1) {
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:760]:
```suggestion
            if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT > -1) {
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 11]
```diff
@@ -611,6 +722,58 @@ public Void call() throws Exception {
...
+
+                int cur_files = new File(file.getParent()).listFiles().length;
+                int limit = Settings.IMP.EXPERIMENTAL.PERPLAYER_FILENUMLIMIT;
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:763]:
```suggestion
                int limit = Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT ;
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 12]
```diff
@@ -611,6 +722,58 @@ public Void call() throws Exception {
...
+
+                if (cur_files >= limit) {
+
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:766]:
```suggestion
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 13]
```diff
@@ -635,8 +798,57 @@ public Void call() throws Exception {
...
+                    if (check_filesize) {
+
+                        double cur_kb = filesize_kb + directorysize_kb;
+                        int allocated_kb = Settings.IMP.EXPERIMENTAL.PERPLAYER_FILESIZELIMIT;
+
+                        if (overwrite) {
+                            cur_kb -= overwrite_old_kb;
+                        }
+
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:817]:
```suggestion
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 14]
```diff
@@ -635,8 +798,57 @@ public Void call() throws Exception {
...
+
+                        if ((cur_kb) > allocated_kb) {
+
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:820]:
```suggestion
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 15]
```diff
@@ -635,8 +798,57 @@ public Void call() throws Exception {
...
+                            file.delete();
+                            TextComponent no_kb_err = TextComponent.of( //TODO - to be moved into captions/translatablecomponents
+                                "You're about to be at " + String.format("%.1f", cur_kb) + "kb of schematics. ("
+                                    + String.format("%dkb", allocated_kb) + " available) Delete some first to save this one!",
+                                TextColor.RED);
+                            actor.printError(no_kb_err);
+
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:827]:
```suggestion
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 16]
```diff
@@ -635,8 +798,57 @@ public Void call() throws Exception {
...
+                        if (overwrite) {
+                            new File(cur_filepath).delete();
+                            file.renameTo(new File(cur_filepath));
+                        }
+                        TextComponent kb_left_notif = TextComponent.of( //TODO - to be moved into captions/translatablecomponents
+                            "You have " + String.format("%.1f", (allocated_kb - cur_kb)) + "kb left for schematics.", TextColor.GRAY);
+                        actor.print(kb_left_notif);
+                    }
+
+                    if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PERPLAYER_FILENUMLIMIT > -1) {
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:841]:
```suggestion
                    if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT > -1) {
```

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 17]
```diff
@@ -635,8 +798,57 @@ public Void call() throws Exception {
...
+                        int cur_files = new File(file.getParent()).listFiles().length;
+
+                        TextComponent slots_left_notif = TextComponent.of( //TODO - to be moved into captions/translatablecomponents
+                            "You have " + (Settings.IMP.EXPERIMENTAL.PERPLAYER_FILENUMLIMIT - cur_files)
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:845]:
```suggestion
                            "You have " + (Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT - cur_files)
```

**Review by N0tMyFaultOG — CHANGES_REQUESTED** [review]
Please remove the underscores in your method names.

**Comment by N0tMyFaultOG** [comment]
Please address the changes requested before requesting a new review.

**Comment by Ivan8or** [comment]
> Please address the changes requested before requesting a new review.

sorry, i got ahead of myself

**Comment by Ivan8or** [comment]
everything should be good now, fingers crossed

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 18]
```diff
@@ -355,19 +383,18 @@ public void save(Actor actor, LocalSession session,
...
             }
         }
 
         ClipboardHolder holder = session.getClipboard();
 
         SchematicSaveTask task = new SchematicSaveTask(actor, f, format, holder, overwrite);
         AsyncCommandBuilder.wrap(task, actor)
-                .registerWithSupervisor(worldEdit.getSupervisor(), "Saving schematic " + filename)
-                .sendMessageAfterDelay(TranslatableComponent.of("worldedit.schematic.save.saving"))
-                .onSuccess(filename + " saved" + (overwrite ? " (overwriting previous file)." : "."), null)
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:368]: What happened to the onSuccess part?

**Comment by N0tMyFaultOG** [comment]
> everything should be good now, fingers crossed

Looks good. Adjust the import outlined plus use proper camel casing for your method names.

**Review by SirYwell — CHANGES_REQUESTED** [review]
(local) variables should be camelCase, not snake_case

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java (hunk 18 above)
Comment (Ivan8or) [reply worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:368]: > What happened to the onSuccess part?

originally the player would get two notifications for a successful save, one coming from where you pointed out,
`        AsyncCommandBuilder.wrap(task, actor)
            .registerWithSupervisor(worldEdit.getSupervisor(), "Saving schematic " + filename)
            .sendMessageAfterDelay(TranslatableComponent.of("worldedit.schematic.save.saving"))
            .onFailure("Failed to save schematic", worldEdit.getPlatformManager().getPlatformCommandManager().getExceptionConverter())
            .buildAndExec(worldEdit.getExecutorService());
`
and the other coming from
`log.info(actor.getName() + " saved " + file.getCanonicalPath());
 actor.print(Caption.of("fawe.worldedit.schematic.schematic.saved", SCHEMATIC_NAME));`


I thought telling the player they had succeeded two times was redundant, so I dropped one of them - though looking back, maybe I should have removed the second one instead. Should I do that?

Comment (SirYwell) [reply worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:368]: Yeah it's probably better to remove the second one but keep the log output

Comment (Ivan8or) [reply worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:368]: will do!

**Comment by Ivan8or** [comment]
All my variables are camel case now, and I got rid of the unused imports 👍
I also moved my error messages on running out of schematic slots/space into WorldEditExceptions in order to make better use of 
`.onFailure("Failed to save schematic", worldEdit.getPlatformManager().getPlatformCommandManager().getExceptionConverter())`

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java [hunk 19]
```diff
@@ -611,6 +719,56 @@ public Void call() throws Exception {
...
+
+
+            if (Settings.IMP.PATHS.PER_PLAYER_SCHEMATICS && Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT > -1) {
+
+                int curFilesAmount = new File(file.getParent()).listFiles().length;
+                int limit = Settings.IMP.EXPERIMENTAL.PER_PLAYER_FILE_NUM_LIMIT;
+
+                if (curFilesAmount >= limit) {
+                    TextComponent noSlotsErr = TextComponent.of( //TODO - to be moved into captions/translatablecomponents
+                        "You have " + String.format("You have " + curFilesAmount + "/" + limit + " saved schematics. Delete some to save this one!",
```
Comment (N0tMyFaultOG) [inline worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:764]: "You have" + "You have".
Plus this sounds a bit misleading. I'd leave a note about the schematic file size limit.

**Review by N0tMyFaultOG — CHANGES_REQUESTED** [review]

File: worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java (hunk 19 above)
Comment (Ivan8or) [reply worldedit-core/src/main/java/com/sk89q/worldedit/command/SchematicCommands.java:764]: > "You have" + "You have".
good catch ;-;

> Plus this sounds a bit misleading. I'd leave a note about the schematic file size limit.
misleading how? I'm not sure I understand

**Review by N0tMyFaultOG — APPROVED** [review]

**Review by SirYwell — CHANGES_REQUESTED** [review]
Currently, there's a logic flaw: You can save schematics in sub-directories using `/schem save <dir>/<filename>`. Only the folder the schematic is actually saved in is taken into account when calculating the size.

From looking at the code, we'd probably need to clean up the whole per player schematic stuff. Maybe it's even better to use the current class of WE and re-implement our additional features then. Not sure what the others think about that and if you @Ivan8or would like to do that, but I'd really like a clean up of that class.

If you don't want to clean it up, we can also merge this one and someone (maybe myself) will do that at some point.

**Comment by Ivan8or** [comment]
> Currently, there's a logic flaw: You can save schematics in sub-directories using `/schem save <dir>/<filename>`. Only the folder the schematic is actually saved in is taken into account when calculating the size.

You're very right, I hadn't thought about that - it should be a straightforward fix though, I think there's already a method in that class which recursively finds all the schematic files in a directory

> From looking at the code, we'd probably need to clean up the whole per player schematic stuff. Maybe it's even better to use the current class of WE and re-implement our additional features then. Not sure what the others think about that and if you @Ivan8or would like to do that, but I'd really like a clean up of that class.

I definitely agree that the per-player schematics need a redesign, and I'm totally happy to help with anything regarding them whenever I have the time 

> If you don't want to clean it up, we can also merge this one and someone (maybe myself) will do that at some point.

I would personally prefer to first merge this in it's current form (after the sub-directory issue is addressed, of course) just to solve the original issue I had with the system asap, and then take a step back and figure out a cleaner way to implement per-player schematics in their entirety

**Review by SirYwell — APPROVED** [review]

**Pull request merged**